go/merkletree/tree_hasher.go

```diff
@@ -0,0 +1,52 @@
+package merkletree
+
+import (
+	"hash"
+)
+
+const (
+	// LeafPrefix is the domain separation prefix for leaf hashes.
+	LeafPrefix = 0
+
+	// NodePrefix is the domain separation prefix for internal tree nodes.
+	NodePrefix = 1
+)
+
+// TreeHasher performs the various hashing operations required when manipulating MerkleTrees.
+type TreeHasher struct {
+	hasher hash.Hash
+}
+
+// NewTreeHasher returns a new TreeHasher based on the passed in hash.
+func NewTreeHasher(h hash.Hash) *TreeHasher {
+	return &TreeHasher{
+		hasher: h,
+	}
+}
+
+// DigestSize returns the size in bytes of the underlying hash.
+func (h TreeHasher) DigestSize() int {
+	return h.hasher.Size()
+}
+
+// HashEmpty returns the hash of the empty string.
+func (h TreeHasher) HashEmpty() []byte {
+	h.hasher.Reset()
+	return h.hasher.Sum([]byte{})
+}
+
+// HashLeaf returns the hash of the passed in leaf, after applying domain separation.
+func (h TreeHasher) HashLeaf(leaf []byte) []byte {
+	h.hasher.Reset()
+	h.hasher.Write([]byte{LeafPrefix})
+	return h.hasher.Sum(leaf)
+
+}
```

Comment:
Can't you do something like this in Go instead? Wouldn't that be more Go-like?

```golang
func (h hash.Hash) HashLeaf(leaf []byte) []byte {
	h.Reset()
	h.Write([]byte{LeafPrefix})
	return h.Sum(leaf)
}
```

And just delete `DigestSize`. And inline the two-line implementation of `HashEmpty` entirely at its one call site?

Or if it's weird to add methods (which it kind of is for these methods in particular, from my perception of how this interface is meant to be used), `HashLeaf`and `HashChildren` could just be plain functions that take a `hash.Hash`?